Conformance 1.2 Test Pack (#848)
finos/FDC3: toolbox/fdc3-conformance/FDC3-1.2-Conformance-Test-Cases.md

-  `UCFilteredContext2`: Perform above test
-  `UCFilteredContext3`: Perform above test, except joining a _different_ channel. Check that you _don't_ receive anything.
-  `UCUnsubscribe`: Perform above test, except that after joining, **A** then `unsubscribe()`s the context listener. Check that **A** _doesn't_ receive anything.
-  `UCFilteredContext4`: Perform above test, except that after joining, **A** changes channel with a further _different_ channel.  Check that **A** _doesn't_ receive anything.

## 3. App Channels 

### App Channels Broadcast (Basic)

| App | Step               | Details                                                                                                                                                        |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | A   | createChannel        |`fdc3.getOrCreateChannel("test-channel")`       |
> | A   | addContextListener |Call `testChannel.addContextListener(null, handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object  |
> | B   | createChannel        | `fdc3. getOrCreateChannel("test-channel")`   |
> | B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
> | A   | Receive Context    | Instrument object matches the one broadcast in 2 above.      |
> 
> -  `ACBasicUsage1` Perform above test 
> -  `ACBasicUsage2` Perform above test, but join channel first and then `testChannel.addContextListener()`
> -  `ACBasicUsage3` Do the app B steps first but in reverse order to populate the channel with context, check that A will receive the context after creating

Reviewer: ```suggestion
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener |Call `testChannel.addContextListener(null, handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object  |
| B   | createChannel        | `const testChannel = fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
| A   | Receive Context    | Instrument object matches the one broadcast in 2 above.      |

-  `ACBasicUsage1` Perform above test 

| App | Step               | Details                                                                                                                                                        |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| B   | createChannel        | `fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | getCurrentContext |Call `testChannel.getCurrentContext()`<br>Check that the Context object returned is of the expected type.  |

-  `ACBasicUsage2` Perform above test.
-  `ACBasicUsage3` Perform above test, but in the final step call `getCurrentContext with a type name. I.e.  `testChannel.getCurrentContext("fdc3.instrument")`.
```

Reviewer: This is actually whats implemented for ACBasicUsage2 and ACBasicUsage3 - but I think they just duplicate ACContextHistoryTyped etc..

Hence, they should probably be removed (from both here and the implementation) after checking that they really are duplicates. @robmoffat

Reviewer 2: @robmoffat can confirm `ACContextHistoryTyped` is a duplicate of `ACBasicUsage3`. The only difference is that ACBasicUsage3 is checking context is received for a single fdc3.instrument type, whereas ACContextHistoryTyped checks for two context types. 

Let me know which one you want to remove and I'll remove it from the code

### App Channels Broadcast (Filtered Context)

> | App | Step               | Details                                                                                                                                                              |
> |-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
> | A   | createChannel        |`fdc3.getOrCreateChannel("test-channel")`       |
> | A   | addContextListener | Call `testChannel.addContextListener("fdc3.instrument", handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object |
> | B   | createChannel        | `fdc3. getOrCreateChannel("test-channel")`   |
> | B   | Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |
> | A   | Receive Context    | Instrument object matches the one broadcast in 2 above.<br>Check that the contact is not received.                                                                   |
> 
> -  `ACFilteredContext1`: Perform above test 
> -  `ACFilteredContext2`: Perform above test, but add listeners for both `fdc3.instrument` and `fdc3.contact` in `addContextListener` step.  Both should be received. 
> -  `ACFilteredContext3`: Perform above test, except creating a _different_ channel. Check that you _don't_ receive anything.
> -  `ACUnsubscribe`: Perform above test, except that after creating the channel **A** then `unsubscribe()`s the channel. Check that **A** _doesn't_ receive anything.
> -  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel.  Check that **A** _doesn't_ receive anything.
> -  `ACFilteredContext5`: Perform above test, except that after creating the channel **A** calls `fdc3.leaveChannel()` and doesn't receive anything.

Reviewer: ```suggestion
| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener | Call `testChannel.addContextListener("fdc3.instrument", handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object |
| B   | createChannel        | `const testChannel = await fdc3. getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |
| A   | Receive Context    | Instrument object matches the one broadcast in 2 above.<br>Check that the contact is not received.                                                                   |

-  `ACFilteredContext1`: Perform above test 
-  `ACFilteredContext2`: Perform above test, but add listeners for both `fdc3.instrument` and `fdc3.contact` in `addContextListener` step.  Both should be received. 
-  `ACFilteredContext3`: Perform above test, except creating a _different_ channel. Check that you _don't_ receive anything.
-  `ACUnsubscribe`: Perform above test, except that after creating the channel **A** then `unsubscribe()`s the listener it added to the channel. Check that **A** _doesn't_ receive anything.
-  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel id.  Check that **A** _doesn't_ receive anything.
```

FYI ACFilteredContext5 is not possible (can't join/leave an app channel)

### App Channel History

| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`fdc3.getOrCreateChannel("test-channel")`       |

> | B   | createChannel        | `fdc3. getOrCreateChannel("test-channel")`   |

Reviewer: ```suggestion
| B   | createChannel        | `fdc3.getOrCreateChannel("test-channel")`   |
```

| B   | Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |

> | A   | Receive Context    | `testChannel.getCurrentContext('fdc.instrument')` returns the last instrument<br>testChannel.getCurrentContext('fdc.contact')` returns the last broadcast contact                                                              |

Reviewer: ```suggestion
| A   | Receive Context    | `testChannel.getCurrentContext('fdc3.instrument')` returns the last instrument<br>testChannel.getCurrentContext('fdc3.contact')` returns the last broadcast contact                                                              |
```

-  `ACContextHistoryTyped`: Perform above test.

> -  `ACContextHistoryMultiple`: **B** Broadcasts multiple history items of both types.  Only the last version of each type is received by **A**.

Reviewer: n.b. implementation of this one is only testing the instrument type - it should be updated to do both types as written here @robmoffat

Reviewer: @Joe-Dunleavy FYI

Reviewer 2: I've fixed this in code. Will create a PR once the comment I left above is resolved

-  `UCFilteredContext1`: Perform above test 

| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | 1. addContextListener | Call `addContextListener (“fdc3.instrument”, handler)`<br>Check listener object returned<br>Check that there is an unsubscribe function on the returned object<br>Call `addContextListener (“fdc3.contact”, handler)`<br>Check listener object returned<br>Check that there is an unsubscribe function on the returned object                                                                                                                                                                                   |

> | A   | 2. joinChannel        | Check that there is an unsubscribe function on the returned object                                                                                                                                                                                                                                                                                |

Reviewer: this should replicate the next step (so that you have a channel to join)

```suggestion
| A   | 2. joinChannel        | `fdc3.getSystemChannels()`<br>Check channels are returned.<br>Call `fdc3.joinChannel()` on first non-global channel. Check that there is an unsubscribe function on the returned object                                                                                                                                                                                                                                                                                |
```

-  `UCFilteredContext3`: Perform above test, except joining a _different_ channel. Check that you _don't_ receive anything.
-  `UCUnsubscribe`: Perform above test, except that after joining, **A** then `unsubscribe()`s the context listener. Check that **A** _doesn't_ receive anything.
-  `UCFilteredContext4`: Perform above test, except that after joining, **A** changes channel with a further _different_ channel.  Check that **A** _doesn't_ receive anything.

## 3. App Channels 

### App Channels Broadcast (Basic)

| App | Step               | Details                                                                                                                                                        |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener |Call `testChannel.addContextListener(null, handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object  |

> | B   | createChannel        | `const testChannel = fdc3. getOrCreateChannel("test-channel")`   |

Reviewer 3: ```suggestion
| B   | createChannel        | `const testChannel = fdc3.getOrCreateChannel("test-channel")`   |
```

| B   | Broadcast          | `testChannel.broadcast(<some instrument>)`   |
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | getCurrentContext |Call `testChannel.getCurrentContext()`<br>Check that the Context object returned is of the expected type.  |

-  `ACBasicUsage2` Perform above test.

### App Channels Broadcast (Filtered Context)

| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | addContextListener | Call `testChannel.addContextListener("fdc3.instrument", handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object |

> | B   | createChannel        | `const testChannel = await fdc3. getOrCreateChannel("test-channel")`   |

Reviewer 3: ```suggestion
| B   | createChannel        | `const testChannel = await fdc3.getOrCreateChannel("test-channel")`   |
```

| B   | 4. Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |
| A   | 5. Receive Context    | Instrument object matches the one broadcast in 4 above.<br>Check that the contact is not received.                                                                   |

-  `ACFilteredContext1`: Perform above test 
-  `ACFilteredContext2`: Perform above test, but add listeners for both `fdc3.instrument` and `fdc3.contact` in `addContextListener` step.  Both should be received. 

> -  `ACFilteredContext3`: Perform above test, except creating a _different_ channel. Check that you _don't_ receive anything.

Reviewer: Small clarification:
```suggestion
-  `ACFilteredContext3`: Perform above test, except creating a _different_ channel in app B. Check that you _don't_ receive anything (as the channels don't match).
```

-  `ACUnsubscribe`: Perform above test, except that after creating the channel **A** then `unsubscribe()`s the listener it added to the channel. Check that **A** _doesn't_ receive anything.

> -  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel id.  Check that **A** _doesn't_ receive anything.

Reviewer: This is wrong - if you create a further channel and still have listeners on the other channel it should still receive stuff - this test is supposed to check that the additional channel does NOT break the first.

```suggestion
-  `ACFilteredContext4`: Perform above test, except that after creating the channel **A** creates another channel with a further _different_ channel id and adds a further context listener to it.  Check that **A** is still able to receive context on the first channel (i.e. it is unaffected by the additional channel) and does *NOT* receive anything on the second channel.
```

@Joe-Dunleavy and @robmoffat please check that this is how the test was implemented.

### App Channel History

| App | Step               | Details                                                                                                                                                              |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | createChannel        |`fdc3.getOrCreateChannel("test-channel")`       |
| B   | createChannel        | `fdc3.getOrCreateChannel("test-channel")`   |
| B   | Broadcast          | `testChannel.broadcast()` the instrument context.<br>`testChannel.broadcast()` a contact context. |

> | A   | Receive Context    | `testChannel.getCurrentContext('fdc3.instrument')` returns the last instrument<br>testChannel.getCurrentContext('fdc3.contact')` returns the last broadcast contact                                                              |

Reviewer 3: ```suggestion
| A   | Receive Context    | `testChannel.getCurrentContext('fdc3.instrument')` returns the last broadcast instrument<br>`testChannel.getCurrentContext('fdc3.contact')` returns the last broadcast contact                                                              |
```

-  `ACContextHistoryLast`: **A** calls `testChannel.getCurrentContext()` retrieves the last broadcast context item

## 4. Open API 

### A Opens B

- `AOpensB1`:  **A** calls `fdc3.open(‘app B Name’)`, check app **B** opens
- `AOpensB2`:  **A** calls `fdc3.open({name: “<app B Name>”})`, check app **B** opens
- `AOpensB3`:  **A** calls `fdc3.open({name: “<app B Name>”, appId: “<app B ID”})`, check app **B** opens

### A Fails To Open B

- `AFailsToOpenB1-3`:  Run the above 4 tests again with a non-existent app name/app id.  Should return “App Not Found” Error from https://fdc3.finos.org/docs/api/ref/Errors#openerror

### A Opens B With Context

| App | Step            | Description                                              |
|-----|-----------------|----------------------------------------------------------|
| A   | 1. Opening App     | various open methods as in `AOpensB1-3` except with a `<context>` argument of type `fdc3.testReceiver` <br>check app opens    |
| B   | 2. Context present | `fdc3.addContextListener()`<br>- receives `<context>` from **A** |

-  `AOpensBWithContext1-3`: Perform above tests

> -  `AOpensBWithSpecificContext`: Perform above but replace **B**s call with `fdc3.addContextListener('fdc3.testReceiver`)`

Reviewer: The other tests in this group have 3 variations. This doesn't appear to (and probably doesn't need to) - so which variation should it be using (doesn't matter, but be specific)?

### Specific Context

| App | Step            | Description                                                                                                                   |
|-----|-----------------|-------------------------------------------------------------------------------------------------------------------------------|
| A   | Opening App     | `fdc3.open(‘app Name’, <contact context>)` <br>check app opens                                                                |
| B   | Context present | fdc3.addContextListener()<br>- receives <context> from A                                                                      |
| A   | Promise         | - receives a rejection from the open promise with “App Timeout’ from <br>https://fdc3.finos.org/docs/api/ref/Errors#openerror |

-  `AOpensBWithWrongContext`: As above

> -  `AOpensBNoListen`: Skip `fdc3.addContextListener() above. 

Reviewer 3: ```suggestion
-  `AOpensBNoListen`: Skip `fdc3.addContextListener()` above. 
```

> -  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  The correct context listener should receive the context, and the promise resolves successfully

Reviewer: ```suggestion
-  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  Check that the specific context listener for `fdc3.instrument` does not receive the context, but that the listener with an unspecified type does, and the promise resolves successfully.
```

> -  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  The correct context listener should receive the context, and the promise completes successfully

Reviewer 3: ```suggestion
-  `AOpensBMultipleListen`:  **B** performs `fdc3.addContextListener('fdc3.instrument') prior to the existing `addContextListener`.  The correct context listener should receive the context, and the promise resolves successfully
```

> -  `AOpensBMalformedContext`: **A** tries to pass malformed context to **B**.  Context listener receives nothing, promise completes successfully.

Reviewer: Describe how its malformed - e.g. no type field.

Also we assume a fourth app **D** that is going to discover the intents in the other 3.

### Find Intent From AppD

-  `IntentAppD`: Calls `fdc3.findIntent(‘aTestingIntent’)`.  Receives promise containing an appIntent with metadata containing `aTestingIntent` and only **A** app metadata.
-  `WrongIntentAppD`: Calls `fdc3.findIntent(‘nonExistentIntent’)`. Rejects with no apps found error https://fdc3.finos.org/docs/api/ref/Errors#resolveerror
-  `IntentAppDRightContext`: Calls `fdc3.findIntent(‘aTestingIntent’, ‘fdc3.testContextX’)`.  Receives promise containing an appIntent with metadata containing `aTestingIntent` and only **A** app metadata.
-  `IntentAppDWrongContext`: Calls `fdc3.findIntent(‘aTestingIntent’, ‘fdc3.testContextY’)`.  Rejects with no apps found error https://fdc3.finos.org/docs/api/ref/Errors#resolveerror
-  `IntentAppDMultiple1`: Calls `fdc3.findIntent(‘sharedTestingIntent1’)`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **A** and **B** app metadata.

> -  `IntentAppDMultiple2`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextX`)`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **A** and **B** app metadata.

Reviewer 3: ```suggestion
-  `IntentAppDMultiple2`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextX')`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **A** and **B** app metadata.
```

> -  `IntentAppDMultiple3`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextY`)`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **B** app metadata.

Reviewer 3: ```suggestion
-  `IntentAppDMultiple3`: Calls `fdc3.findIntent(‘sharedTestingIntent1’, 'testContextY')`.  Receives promise containing an appIntent with metadata containing `sharedTestingIntent` and only **B** app metadata.
```

-  `NoContext`: Call `fdc3.findIntentsByContext()`. Throws error of `NoAppsFound`

### Raise Intent

| App | Step           | Details                                                                                           |
|-----|----------------|---------------------------------------------------------------------------------------------------|
| A   | 1. Raise          | `fdc3.raiseIntent(‘sharedTestingIntent1’, {testContextY})`<br>starts app B.                       |
| B   | 2. Gather Context | `fdc.addIntentListener(‘sharedTestingIntent1’)`<br>Receives testContextY, matching that sent by D |

-  `SingleResolve1`: Perform above test
-  `TargetedResolve1`: Use `fdc3.raiseIntent(‘aTestingIntent’, {testContextX}, <A’s App Name>)` to start app A, otherwise, as above
-  `TargetedResolve2-3` Use the other ways of addressing apps (via ID, metadata) as described in `AOpensB2-3` 
-  `FailedResolve1-3` As above, but use `fdc3.raiseIntent(‘aTestingIntent’, {testContextY}, <A’s App Name>)` and variations.  You will receive `NoAppsFound` Error

> -  `FailedResolve4` As above, but use `fdc3.raiseIntent(‘aTestingIntent’, {testContextX}, <C’s App Name>)`.  You will receive `NoAppsFound` Error

Reviewer: I'd prefer variations to be spelled out here rather than referencing other tests or just 'and variations' - but this is otherwise correct.

- `BasicCL1`: You can call the `fdc3.addContextListener` with the `fdc3.contact` context type. The returned listener object has an `unsubscribe` function.
- `BasicCL2`: You can call the `fdc3addContextListener` with no context type. The returned listener object has an `unsubscribe` function.
- `BasicIL1`: You can call the `fdc3.addIntentListener` on the `DesktopAgent` for an intent, and get back a `Listener` object with `unsubscribe` method.
- `BasicCH1`: A call to `fdc3.getCurrentChannel` on the `DesktopAgent` always returns a promise.
- `BasicCH2`: A call to `fdc3.getCurrentChannel()` returns a promise resolving to _null_ if called prior to any `joinChannel`.
- `BasicGI1`: A call to `fdc3.getInfo()` returns an object with `fdc3Version` and `provider` properties.
- `BasicAC1`: A call to `fdc3.getOrCreateChannel(<name>)` will return an promise resolving to an object matching the `Channel` interface, with properties of `id`, `type`, `broadcast`, `getCurrentContext` and `addContextListener`.
- `BasicUC1`: You can call the `fdc3.getSystemChannels()` function and receive a promise containing an array of more than 1 `Channel` objects, each with `type` and `id` set.

> - `BasicJC1`: You can call `fdc3.joinChannel`, passing in the `id` of one of the system channels.  After the returned promise is resolved, `fdc3.getCurrentChannel` should then return that joined channel.

Reviewer: @robmoffat Kalina's PR is adding a BasicJC2 and BasicJC3 - BasicJC2 makes sense and needs adding here, but we can't do BasicJC3 (for the same reason others don't want to test OpenError.AppTimeout, it isn't clear that it could/should be returned if the channel was not in system channels). In fact *most* implementations of 1.2 allow you to join AppChannels - it was only in 2.0 that we started to clarify the intention that it should only be user channels that you join (and documented the relevant error).

-  `UCUnsubscribe`: Perform above test, except that after joining, **A** then `unsubscribe()`s the context listener. Check that **A** _doesn't_ receive anything.
-  `UCFilteredContext4`: Perform above test, except that after joining, **A** changes channel with a further _different_ channel.  Check that **A** _doesn't_ receive anything.

## 3. App Channels 

### App Channels Broadcast (Basic)

| App | Step               | Details                                                                                                                                                        |
|-----|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| A   | 1. createChannel        |`const testChannel = await fdc3.getOrCreateChannel("test-channel")`       |
| A   | 2. addContextListener |Call `testChannel.addContextListener(null, handler)`<br>Check listener object returned<br>Check that there is an `unsubscribe` function on the returned object  |

> | B   | 3. createChannel        | `const testChannel = fdc3. getOrCreateChannel("test-channel")`   |

Reviewer 4: Let's remove the space after fdc3.
